File: ring_attractor.py
from datetime import datetime
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from utils import circular_mean
from lif_model import LIF


class RingAttractor:
    "A self-contained class for the ring attractor"

    def __init__(self,
                 n=128,
                 noise=2.0e-3,
                 # TODO: pick more stable fixed point weights
                 weights=(0.050, 0.088, 0.050, 0.15),
                 fixed_points_number=0, 
                 time=300,
                 plot=False):

        self.n = n
        self.noise = noise
        self.weights = weights
        self.fp_n = fixed_points_number
        self.time = time
        self.plot = plot

        self.neurons = [LIF(ID=i, angle=360.0/n*i, noise_mean=0, noise_std=self.noise,)
                        for i in range(n)]

        # TODO: move weights, fp_n and noise to the simulate method
        self.fixed_points = self.get_fixed_points()

        self.connect_with_fixed_points()

    def simulate(self):

        mid_point = self.get_mid_point()

        potentials = [[] for _ in range(self.n)]
        for t in range(self.time):
            for neuron in self.neurons:

                self.input_source(mid_point=mid_point, n_of_spikes=5,
                                  weight=self.weights[0], begin_time=0, neuron=neuron, time=t)
                neuron.step()
                potentials[neuron.id].append(neuron.V)

        df, e = self.compute_gain(potentials, mid_point)

        if self.plot:
            self.plot_potentials(df, e)

        return e

    def input_source(self, mid_point, n_of_spikes, weight, begin_time, neuron, time):
        sources = [i for i in range(mid_point - 2, mid_point + 3)]

        if time > begin_time:
            if neuron.id in sources:
                for _ in range(n_of_spikes):
                    neuron.exc_ps_td.append(
                        ((time - begin_time) * 1e-3, weight))

    def connect_with_fixed_points(self):
        for neur in self.neurons:
            if neur.id in self.fixed_points:
                for i in range(5, 12):
                    neur.synapses["inh"][self.neurons[(
                        neur.id + i) % self.n]] = self.weights[3]
                    neur.synapses["inh"][self.neurons[neur.id - i]
                                         ] = self.weights[3]
                for i in range(1, 5):
                    neur.synapses["exc"][self.neurons[(
                        neur.id + i) % self.n]] = self.weights[2]
                    neur.synapses["exc"][self.neurons[neur.id - i]
                                         ] = self.weights[2]

            else:
                for i in range(5, 12):
                    neur.synapses["inh"][self.neurons[(
                        neur.id + i) % self.n]] = self.weights[1]
                    neur.synapses["inh"][self.neurons[neur.id - i]
                                         ] = self.weights[1]
                for i in range(1, 5):
                    neur.synapses["exc"][self.neurons[(
                        neur.id + i) % self.n]] = self.weights[0]
                    neur.synapses["exc"][self.neurons[neur.id - i]
                                         ] = self.weights[0]

    def get_fixed_points(self):
        if self.fp_n == 0:
            return []

        index = np.arange(self.n)
        interval = self.n // self.fp_n

        distances = index % interval
        return np.where(distances < 3)[0]

    def get_mid_point(self):
        if self.fp_n <= 1:
            return self.n // 2


        free_points = set(np.arange(self.n)) - set(self.fixed_points)
        median = [*free_points][len(free_points) // 4]


        high = self.fixed_points[self.fixed_points > median][0]
        low = self.fixed_points[self.fixed_points < median][-1]
        mid_point = (high + low) // 2

        return mid_point

    def compute_gain(self, potentials, mid_point):
        df = pd.DataFrame(potentials)
        df.index = [self.neurons[i].angle for i in df.index]
        spikes = df == 0.0

        # sparse = spikes.astype(pd.SparseDtype())
        # coo = sparse.sparse.to_coo()
        # coo.eliminate_zeros()

        # regr = LinearRegression(fit_intercept=True)
        
        # x = coo.col.reshape(-1, 1)
        # y = coo.row.reshape(-1, 1)

        # regr.fit(x, y)
        # error = np.abs(regr.coef_[0][0])

        

        spikes = spikes.iloc[:, -30:]
        spikes = spikes.astype(int)
        spikes = spikes.apply(lambda x: x * x.index)
        spikes = spikes.replace(0, np.nan)

        means = spikes.apply(circular_mean)
        total_mean = circular_mean(means)
        error = np.abs(self.neurons[mid_point].angle - total_mean)

        print("total_mean", total_mean)
        print(self.neurons[mid_point].angle)

        return df, error


    def plot_potentials(self, df, error):
        _, ax = plt.subplots(figsize=(10, 10))
        sns.heatmap(df, vmin=-0.08, vmax=0.0, cmap="viridis", xticklabels=int(self.time/10),
                    yticklabels=5, cbar_kws={'label': "Membrane Potential (V)"}, ax=ax)
        plt.xlabel("Time (ms)")
        plt.ylabel("Orientation of neuron")
        plt.subplots_adjust(left=0.07, bottom=0.07, right=0.97, top=0.89)

        labels = [item.get_text() for item in ax.get_yticklabels()]

        # for i, l in enumerate(labels):
        #     if int(l) in fixed_points:
        #         labels[i] = labels[i] + '\nFP'

        ax.set_yticklabels(labels)

        ax.set_title("Number of fixed points: {}\nNoise: {:.3e}\nWeights: {}\nError: {:.3f}".format(
            self.fp_n, self.noise, self.weights, error * self.time))

        plt.savefig(
            f"images/{datetime.now().strftime('%d-%m-%Y, %H:%M:%S')}.png")
        plt.show()

if __name__ == "__main__":

    # np.random.seed(42)
    ring = RingAttractor(n=64, noise=3e-3, weights=(0.050, 0.100, 0.050, 0.250), fixed_points_number=2, time=100, plot=True)
    error = ring.simulate()

File: ring-attractor.jl
module RingAttractor

using Distributions
using Random

const ms = 1e-3
const mV = 1e-3
const nF = 1e-9
const dt = 1e-3

const Vₜ = -48 * mV
const Vᵣ = -80 * mV
const Cₘ = 1e-9
const Eᵢ = -70 * mV
const Eₑ = 0.
const Eₘ = -70 * mV
const Eₗ = -70 * mV
const τᵣ = 2 * ms
const τₛ = 5 * ms
const τₘ = 5 * ms
const kₛ = 1/(5 * ms * exp(-1.))
const Eₗₜ = Eₗ/τₛ

const Nₑ = 5
const Nᵢ = 7
const fp_w = 2


mutable struct Ring <: Function
	N::Int32
	time::Int32
	t::Int32

	noise::Float64
	fps::Tuple # TODO take fpn and autogenerate fps
	seed::Int32

	wₑ::Float64
	wᵢ::Float64
	wₑᶠ::Float64
	wᵢᶠ::Float64
	Wₑ::Array{Float64, 2}
	Wᵢ::Array{Float64, 2}


	S::Array{Bool, 2}
	V::Array{Float64, 2}
	Z::Array{Float64, 2}

	Δs::Array{Float64, 2}
	k::Array{Float64, 1}


	function Ring(;N=64, time=10000, noise=5e-4, fps=(), seed=0, wₑ=0.05, wᵢ=0.10, wₑᶠ=0.05, wᵢᶠ=0.25)
		self = new()

		self.N = N
		self.time = time
		self.noise = noise
		self.fps = fps
		self.seed = seed
		if seed != 0
			Random.seed!(seed)
		end

		self.wₑ = wₑ
		self.wᵢ = wᵢ
		self.wₑᶠ = wₑᶠ
		self.wᵢᶠ = wᵢᶠ

		self.Wₑ = zeros(N, N)
		self.Wᵢ = zeros(N, N)
		self.V = zeros(N, time)
		self.Z = zeros(N, time)
		self.S = falses(N, time)
		self.Δs = zeros(N, 3)
		self.k = zeros(N)
		self.t = 0

		init!(self)

		return self
	end
end


function (r::Ring)()
	init!(r)

	for r.t = 1:r.time-1
		idx = r.t % 3 + 1
		r.S[:, r.t] .= view(r.V, :, r.t) .== 0.

		r.k .= @views r.Δs[:, idx] .* exp.(-r.Δs[:, idx] ./ τₛ)

		r.V[:, r.t+1] .= @views r.V[:, r.t] .+ ((Eₗₜ .- r.V[:, r.t] ./ τₘ ) .- (r.V[:, r.t].-Eₑ) .* (r.Wₑ' * r.k) .- (r.V[:, r.t].-Eᵢ) .* (r.Wᵢ' * r.k)) .* dt
		r.V[:, r.t+1] .+= @view r.Z[:, r.t+1]

		r.V[view(r.V,:, r.t) .> Vₜ, r.t+1] .= 0.
		r.V[view(r.S, :, r.t), r.t+1] .= Vᵣ

		r.Δs .+= dt
		r.Δs[view(r.S, :, r.t), idx] .= 0.
	end
end


function setweights!(r::Ring)
	r.Wₑ .= reshape(Float64[min(r.N - abs(i-k), abs(i-k)) for i in 1:r.N for k in 1:r.N], (r.N,r.N))
	r.Wᵢ .= deepcopy(r.Wₑ)

	replace!(x -> 0., view(r.Wₑ, r.Wₑ .> Nₑ))
	replace!(x -> r.wₑ, view(r.Wₑ, r.Wₑ .> 0.))

	replace!(x -> 0., view(r.Wᵢ, Nₑ .>= r.Wᵢ))
	replace!(x -> 0., view(r.Wᵢ, r.Wᵢ .> Nₑ + Nᵢ))
	replace!(x -> r.wᵢ,  view(r.Wᵢ, r.Wᵢ .> 0.))

    for fp in r.fps
		r = (fp:fp + 2) .- 1
		replace!(view(r.Wₑ, r, :), r.wₑ => r.wₑᶠ)
		replace!(view(r.Wᵢ, r, :), r.wᵢ => r.wᵢᶠ)
    end

	r.Wₑ .*= kₛ * 1e-6 / Cₘ
	r.Wᵢ .*= kₛ * 1e-6 / Cₘ
end

function init!(r::Ring)
	r.Δs = fill(0.2, r.N, 3)
	r.Z = rand(Normal(0., r.noise), r.N, r.time)
	replace!(view(r.V, :, 1), 0. => Vᵣ)
	replace!(view(r.V, (r.N÷2:r.N÷2+5).-2, 1), Vᵣ => 0.)
	setweights!(r)
end

export Ring

end
